web/book/src/internals/modules.md

```diff
@@ -0,0 +1,325 @@
+    # _my_project.prql
+
+    module sales {
+        # sales.prql
+    }
+
+    module projections {
+        # projections/_projections.prql
+
+        module year_2023 {
+            # projections/year_2023.prql
+        }
+
+        module year_2024 {
+            # projections/year_2024.prql
+        }
+    }
+}
+```
+
+If module `my_project.sales` wants to add a submodule `util`, it has to be
+converted to a directory modules. This means that it has to be moved to
+`sales/_sales.prql`. The new module would reside in `sales/util.prql`.
+
+The annotated layout is not the only possible layout for this module structure,
+since any of the modules `sales`, `year_2023` or `year_2024` could be converted
+into a directory module with zero sub-modules.
+
+Point 4 of design goals means that each declaration within a project has a
+single fully-qualified name within this project. This is ensured by strict rules
+regarding importing files and the fact that the module structure is a tree.
+
+## Declaration order
+
+The order of declarations in a module holds no semantic value, except the "last
+`main` variable".
+
+References between modules can be cyclic.
+
+```
+module mod_a {
+    let decl_a_1 = ...
+    let decl_a_2 = (from mod_b.decl_b | take 10)
+}
+module mod_b {
+    let decl_b = (from mod_a.decl_a | take 10)
+}
+```
+
+References between variable declarations cannot be cyclic.
+
+```
+let decl_a = (from decl_b)
+let decl_b = (from decl_a) # error: cyclic reference
+```
+
+```
+module mod_a {
+    let decl_a = (from mod_b.decl_b)
+}
+module mod_b {
+    let decl_b = (from mod_a.decl_a) # error: cyclic reference
+}
+```
+
+## Compiler interface
+
+`prql-compiler` provides two interfaces for compiling files.
+
+**Multi-file interface** requires three arguments:
+
+- path to the file containing the module which is the root of the module
+  structure,
+- identifier of the pipeline that should be compiled to RQ (this can also be an
+  identifier of a module that has a `main` pipeline) and,
+- a "file loader", which can load files on-demand.
+
+The path to the root module can be automatically detected by searching for
+`.prql` files starting with `_` in the current working directory.
+
+Example prqlc usage:
+
+```
+$ prqlc compile _project.prql sales.projections.orders_2024
+$ prqlc compile sales.projections.orders_2024
+```
+
+**Single-file interface** requires a single argument; the PRQL source. Any
+attempts to load modules in this mode result in compilation errors. This
+interface is needed, for example, when integrating the compiler with a database
+connector (i.e. JDBC) where no other files can be loaded.
+
+## Built-in module structure
+
+> Work In Progress
+
+```
+# root module of every project
+module project {
+	module std {
+		func sum a -> ...
+		func mean a -> ...
+	}
+
+	module default_db {
+		# all inferred tables and defined CTEs
+	}
+
+	let main = (
+		from t = tracks
+		select [track_id, title]
+	)
+}
+```
+
+## Example
```

Comment:
@vanillajonathan starting a new thread for the "import parent" discussion:

> > Surely this would be super restrictive — any sort of utils package wouldn't be possible, unless every file was in the same path? Are there any other languages that do this? (Or am I misunderstanding the point?)

> C# and Java imports using namespaces so nothing can be from a parent directory. PHP lets you [include](https://www.php.net/manual/en/function.include.php) from any file path including relative paths which may be parent directories. JavaScript (Node.js) lets you import from parent directories. I am not sure about Python, but I don't think it is common. I also worry about the security impact of importing from parent directories.

> In the PHP world when using Composer it is common that imports from your own code is in your your_project/src/ directory, and third-party code from a your_project/vendor/ directory.

Ah, so I think that's where the misunderstanding is — I agree it might be dangerous to allow importing from a parent outside of the package.

Python allows for importing parents within the package — e.g. `from ... import utils`, because the path is relative to the current file. IIUC Java allows for `import utils` in a file `src/foo/bar/baz.java` from a file `src/utils.java`, because the path is relative to the package root.

So given this, I agree possibly we should downweight being able to import from a parent outside the package, which might require a definition of what a package is. (Though I'm not sure, let's see how the other discussions shake out; possibly it's a nice feature to be less tethered to the concept of a package)